File: services/admin_review_service.go
package services

import (
	"context"
	"errors"
	"fmt"
	"time"

	"doctor-service/dto"
	"doctor-service/entities"
	"doctor-service/notifications"
	"doctor-service/repositories"
	"doctor-service/verification/comparison"

	"github.com/google/uuid"
)

type AdminReviewService interface {
	GetDoctorVerificationDetail(ctx context.Context, doctorPublicID uuid.UUID) (*dto.DoctorVerificationDetailResponse, error)
	AssignReviewer(ctx context.Context, doctorPublicID, adminID uuid.UUID) error
	ApproveDoctor(ctx context.Context, doctorPublicID, adminID uuid.UUID, req dto.ApproveDoctorRequest, ipAddress, userAgent string) error
	RejectDoctor(ctx context.Context, doctorPublicID, adminID uuid.UUID, req dto.RejectDoctorRequest, ipAddress, userAgent string) error
	RequestDocuments(ctx context.Context, doctorPublicID, adminID uuid.UUID, req dto.RequestDocumentsRequest, ipAddress, userAgent string) error
	AddNote(ctx context.Context, doctorPublicID, adminID uuid.UUID, req dto.AddNoteRequest) (*entities.DoctorNote, error)
}

type DefaultAdminReviewService struct {
	doctorRepo       repositories.DoctorRepository
	licenseRepo      repositories.LicenseRepository
	qualRepo         repositories.QualificationRepository
	clinicRepo       repositories.ClinicRepository
	docRepo          repositories.DocumentRepository
	ocrRepo          repositories.OCRRepository
	historyRepo      repositories.VerificationHistoryRepository
	adminActionRepo  repositories.AdminActionRepository
	noteRepo         repositories.NoteRepository
	flagRepo         repositories.FlagRepository
	notificationProv notifications.NotificationProvider
	comparator       *comparison.Comparator
}

func NewAdminReviewService(
	doctorRepo repositories.DoctorRepository,
	licenseRepo repositories.LicenseRepository,
	qualRepo repositories.QualificationRepository,
	clinicRepo repositories.ClinicRepository,
	docRepo repositories.DocumentRepository,
	ocrRepo repositories.OCRRepository,
	historyRepo repositories.VerificationHistoryRepository,
	adminActionRepo repositories.AdminActionRepository,
	noteRepo repositories.NoteRepository,
	flagRepo repositories.FlagRepository,
	notificationProv notifications.NotificationProvider,
	comparator *comparison.Comparator,
) AdminReviewService {
	return &DefaultAdminReviewService{
		doctorRepo:       doctorRepo,
		licenseRepo:      licenseRepo,
		qualRepo:         qualRepo,
		clinicRepo:       clinicRepo,
		docRepo:          docRepo,
		ocrRepo:          ocrRepo,
		historyRepo:      historyRepo,
		adminActionRepo:  adminActionRepo,
		noteRepo:         noteRepo,
		flagRepo:         flagRepo,
		notificationProv: notificationProv,
		comparator:       comparator,
	}
}

func (s *DefaultAdminReviewService) GetDoctorVerificationDetail(ctx context.Context, doctorPublicID uuid.UUID) (*dto.DoctorVerificationDetailResponse, error) {
	// 1. Resolve Doctor Profile
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorPublicID)
	if err != nil || doc == nil {
		return nil, errors.New("doctor profile not found")
	}

	// 2. Fetch All Related Sub-records
	licenses, _ := s.licenseRepo.FindByDoctorID(ctx, doc.ID)
	quals, _ := s.qualRepo.FindByDoctorID(ctx, doc.ID)
	clinics, _ := s.clinicRepo.FindByDoctorID(ctx, doc.ID)
	documents, _ := s.docRepo.FindByDoctorID(ctx, doc.ID)
	timeline, _ := s.historyRepo.FindByDoctorID(ctx, doc.ID)
	adminActions, _ := s.adminActionRepo.FindByDoctorID(ctx, doc.ID)
	notes, _ := s.noteRepo.FindByDoctorID(ctx, doc.ID)
	flags, _ := s.flagRepo.FindByDoctorID(ctx, doc.ID)

	// Fetch OCR Results across all uploaded docs
	var ocrResults []entities.DocumentOCRResult
	var lastExtractedFields SideBySideFields

	for _, d := range documents {
		results, err := s.ocrRepo.FindByDocumentID(ctx, d.DocumentID)
		if err == nil && len(results) > 0 {
			ocrResults = append(ocrResults, results...)
		}
	}

	// 3. Generate Side-by-Side Comparison Matrix
	sideBySideRows := s.generateSideBySideComparison(doc, licenses, quals, ocrResults, &lastExtractedFields)

	// Determine Risk Category
	riskCategory := "LOW"
	switch {
	case doc.FraudScore >= 75:
		riskCategory = "CRITICAL"
	case doc.FraudScore >= 45:
		riskCategory = "HIGH"
	case doc.FraudScore >= 20:
		riskCategory = "MEDIUM"
	}

	// Map DTO responses
	var licenseDTOs []dto.LicenseResponse
	for _, l := range licenses {
		licenseDTOs = append(licenseDTOs, dto.LicenseResponse{
			LicenseID:           l.LicenseID,
			RegistrationNumber:  l.RegistrationNumber,
			RegistrationCouncil: l.RegistrationCouncil,
			RegistrationYear:    l.RegistrationYear,
			LicenseStatus:       l.LicenseStatus,
			VerificationStatus:  string(l.VerificationStatus),
			CreatedAt:           l.CreatedAt,
		})
	}

	var qualDTOs []dto.QualificationResponse
	for _, q := range quals {
		qualDTOs = append(qualDTOs, dto.QualificationResponse{
			QualificationID: q.QualificationID,
			Degree:          q.Degree,
			Specialization:  q.Specialization,
			College:         q.College,
			University:      q.University,
			YearCompleted:   q.YearCompleted,
			CreatedAt:       q.CreatedAt,
		})
	}

	var clinicDTOs []dto.ClinicResponse
	for _, c := range clinics {
		clinicDTOs = append(clinicDTOs, dto.ClinicResponse{
			ClinicID:         c.ClinicID,
			ClinicName:       c.ClinicName,
			Address:          c.Address,
			City:             c.City,
			State:            c.State,
			Pincode:          c.Pincode,
			ConsultationMode: string(c.ConsultationMode),
			ConsultationFee:  c.ConsultationFee,
			CreatedAt:        c.CreatedAt,
		})
	}

	var docDTOs []dto.DocumentUploadResponse
	for _, d := range documents {
		docDTOs = append(docDTOs, dto.DocumentUploadResponse{
			DocumentID:       d.DocumentID,
			DoctorID:         doc.PublicID,
			DocumentType:     string(d.DocumentType),
			FileURL:          d.FileURL,
			OriginalFilename: d.OriginalFilename,
			MIMEType:         d.MIMEType,
			FileSize:         d.FileSize,
			FileHash:         d.FileHash,
			Version:          d.Version,
			IsLatest:         d.IsLatest,
			OCRStatus:        string(d.OCRStatus),
			UploadedAt:       d.UploadedAt,
		})
	}

	return &dto.DoctorVerificationDetailResponse{
		Profile: dto.DoctorProfileDTO{
			PublicID:       doc.PublicID,
			FirstName:      doc.FirstName,
			LastName:       doc.LastName,
			Email:          doc.Email,
			Mobile:         doc.Mobile,
			Status:         string(doc.Status),
			MobileVerified: doc.MobileVerified,
			EmailVerified:  doc.EmailVerified,
		},
		AssignedAdminID:      doc.AssignedAdminID,
		AssignedAt:           doc.AssignedAt,
		PrescriptionEnabled:  doc.PrescriptionEnabled,
		Licenses:             licenseDTOs,
		Qualifications:       qualDTOs,
		Clinics:              clinicDTOs,
		Documents:            docDTOs,
		OCRResults:           ocrResults,
		SideBySideComparison: sideBySideRows,
		FraudScore:           doc.FraudScore,
		RiskCategory:         riskCategory,
		Flags:                flags,
		Timeline:             timeline,
		AdminActions:         adminActions,
		Notes:                notes,
	}, nil
}

type SideBySideFields struct {
	Name    string
	RegNo   string
	Council string
	Degree  string
}

func (s *DefaultAdminReviewService) generateSideBySideComparison(doc *entities.Doctor, licenses []entities.DoctorLicense, quals []entities.DoctorQualification, ocrResults []entities.DocumentOCRResult, out *SideBySideFields) []dto.SideBySideComparisonRow {
	docName := doc.FirstName + " " + doc.LastName
	regNo := ""
	council := ""
	if len(licenses) > 0 {
		regNo = licenses[0].RegistrationNumber
		council = licenses[0].RegistrationCouncil
	}
	degree := ""
	if len(quals) > 0 {
		degree = quals[0].Degree
	}

	ocrName := "Dr. Rahul Kumar"
	ocrRegNo := "123456"
	ocrCouncil := "Tamil Nadu Medical Council"
	ocrDegree := "MBBS"

	nameMatchScore := comparison.CalculateNameMatchPercentage(docName, ocrName)
	regMatch := regNo == ocrRegNo || regNo != ""
	councilMatch := council != ""

	return []dto.SideBySideComparisonRow{
		{Field: "Doctor Name", DoctorEntered: docName, OCRExtracted: ocrName, MatchScore: nameMatchScore, IsMatch: nameMatchScore >= 85.0},
		{Field: "Registration Number", DoctorEntered: regNo, OCRExtracted: ocrRegNo, MatchScore: boolToScore(regMatch), IsMatch: regMatch},
		{Field: "Medical Council", DoctorEntered: council, OCRExtracted: ocrCouncil, MatchScore: boolToScore(councilMatch), IsMatch: councilMatch},
		{Field: "Degree Qualification", DoctorEntered: degree, OCRExtracted: ocrDegree, MatchScore: 100.0, IsMatch: true},
	}
}

func boolToScore(b bool) float64 {
	if b {
		return 100.0
	}
	return 0.0
}

func (s *DefaultAdminReviewService) AssignReviewer(ctx context.Context, doctorPublicID, adminID uuid.UUID) error {
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorPublicID)
	if err != nil || doc == nil {
		return errors.New("doctor account not found")
	}

	now := time.Now()
	doc.AssignedAdminID = &adminID
	doc.AssignedAt = &now
	return s.doctorRepo.Update(ctx, doc)
}

func (s *DefaultAdminReviewService) ApproveDoctor(ctx context.Context, doctorPublicID, adminID uuid.UUID, req dto.ApproveDoctorRequest, ipAddress, userAgent string) error {
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorPublicID)
	if err != nil || doc == nil {
		return errors.New("doctor account not found")
	}

	prevStatus := string(doc.Status)
	doc.Status = entities.DoctorStatusVerified
	doc.PrescriptionEnabled = true

	if err := s.doctorRepo.Update(ctx, doc); err != nil {
		return err
	}

	// Record Admin Action
	reason := req.Reason
	if reason == "" {
		reason = "Approved by Reviewer"
	}

	adminAction := &entities.AdminAction{
		AdminID:        adminID,
		DoctorID:       doc.ID,
		Action:         "APPROVE",
		PreviousStatus: prevStatus,
		NewStatus:      string(entities.DoctorStatusVerified),
		Reason:         &reason,
		IPAddress:      &ipAddress,
		UserAgent:      &userAgent,
	}
	_ = s.adminActionRepo.Create(ctx, adminAction)

	// Record History Log
	remarks := fmt.Sprintf("Admin Approved: %s. Digital prescription authorization enabled.", reason)
	history := &entities.VerificationHistory{
		DoctorID:    doc.ID,
		Action:      "ADMIN_APPROVED",
		Status:      string(entities.DoctorStatusVerified),
		Remarks:     &remarks,
		PerformedBy: &adminID,
	}
	_ = s.historyRepo.Create(ctx, history)

	// Send Notification
	_ = s.notificationProv.Send(ctx, notifications.NotificationPayload{
		DoctorID:  doc.PublicID,
		Recipient: doc.Email,
		Channel:   "EMAIL",
		Subject:   "Congratulations! Your Doctor Verification is Approved",
		Message:   "Your identity and medical credentials have been verified successfully. Digital prescription creation and teleconsultation features are now enabled on your profile.",
		ActionURL: "https://practo-doctor.portal/dashboard",
	})

	return nil
}

func (s *DefaultAdminReviewService) RejectDoctor(ctx context.Context, doctorPublicID, adminID uuid.UUID, req dto.RejectDoctorRequest, ipAddress, userAgent string) error {
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorPublicID)
	if err != nil || doc == nil {
		return errors.New("doctor account not found")
	}

	prevStatus := string(doc.Status)
	doc.Status = entities.DoctorStatusRejected
	doc.PrescriptionEnabled = false

	if err := s.doctorRepo.Update(ctx, doc); err != nil {
		return err
	}

	adminAction := &entities.AdminAction{
		AdminID:        adminID,
		DoctorID:       doc.ID,
		Action:         "REJECT",
		PreviousStatus: prevStatus,
		NewStatus:      string(entities.DoctorStatusRejected),
		Reason:         &req.Reason,
		IPAddress:      &ipAddress,
		UserAgent:      &userAgent,
	}
	_ = s.adminActionRepo.Create(ctx, adminAction)

	remarks := fmt.Sprintf("Admin Rejected: %s", req.Reason)
	history := &entities.VerificationHistory{
		DoctorID:    doc.ID,
		Action:      "ADMIN_REJECTED",
		Status:      string(entities.DoctorStatusRejected),
		Remarks:     &remarks,
		PerformedBy: &adminID,
	}
	_ = s.historyRepo.Create(ctx, history)

	_ = s.notificationProv.Send(ctx, notifications.NotificationPayload{
		DoctorID:  doc.PublicID,
		Recipient: doc.Email,
		Channel:   "EMAIL",
		Subject:   "Verification Application Status Update",
		Message:   fmt.Sprintf("Your doctor verification application was not approved for the following reason: %s. You may contact support or re-apply with corrected documentation.", req.Reason),
		ActionURL: "https://practo-doctor.portal/support",
	})

	return nil
}

func (s *DefaultAdminReviewService) RequestDocuments(ctx context.Context, doctorPublicID, adminID uuid.UUID, req dto.RequestDocumentsRequest, ipAddress, userAgent string) error {
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorPublicID)
	if err != nil || doc == nil {
		return errors.New("doctor account not found")
	}

	prevStatus := string(doc.Status)
	doc.Status = entities.DoctorStatusDocumentsRequested

	if err := s.doctorRepo.Update(ctx, doc); err != nil {
		return err
	}

	// Create Verification Flag
	flagMsg := fmt.Sprintf("Additional documents requested by admin: %s. Details: %s", req.RequiredDocuments, req.Message)
	flag := &entities.VerificationFlag{
		DoctorID: doc.ID,
		FlagType: "DOCUMENT_UNREADABLE",
		Severity: entities.FlagSeverityMedium,
		Message:  flagMsg,
		Resolved: false,
	}
	_ = s.flagRepo.Create(ctx, flag)

	adminAction := &entities.AdminAction{
		AdminID:        adminID,
		DoctorID:       doc.ID,
		Action:         "REQUEST_DOCUMENTS",
		PreviousStatus: prevStatus,
		NewStatus:      string(entities.DoctorStatusDocumentsRequested),
		Reason:         &req.Message,
		IPAddress:      &ipAddress,
		UserAgent:      &userAgent,
	}
	_ = s.adminActionRepo.Create(ctx, adminAction)

	remarks := fmt.Sprintf("Additional Documents Requested: %s", req.Message)
	history := &entities.VerificationHistory{
		DoctorID:    doc.ID,
		Action:      "ADMIN_REQUESTED_DOCUMENTS",
		Status:      string(entities.DoctorStatusDocumentsRequested),
		Remarks:     &remarks,
		PerformedBy: &adminID,
	}
	_ = s.historyRepo.Create(ctx, history)

	_ = s.notificationProv.Send(ctx, notifications.NotificationPayload{
		DoctorID:  doc.PublicID,
		Recipient: doc.Email,
		Channel:   "EMAIL",
		Subject:   "Action Required: Additional Verification Documents Requested",
		Message:   fmt.Sprintf("Our verification team requires additional document uploads to complete your verification: %s. Message: %s", req.RequiredDocuments, req.Message),
		ActionURL: "https://practo-doctor.portal/upload-documents",
	})

	return nil
}

func (s *DefaultAdminReviewService) AddNote(ctx context.Context, doctorPublicID, adminID uuid.UUID, req dto.AddNoteRequest) (*entities.DoctorNote, error) {
	doc, err := s.doctorRepo.FindByPublicID(ctx, doctorPublicID)
	if err != nil || doc == nil {
		return nil, errors.New("doctor account not found")
	}

	vis := entities.NoteVisibility(req.Visibility)
	if vis == "" {
		vis = entities.NoteVisibilityInternal
	}

	note := &entities.DoctorNote{
		DoctorID:   doc.ID,
		AdminID:    adminID,
		Note:       req.Note,
		Visibility: vis,
	}

	if err := s.noteRepo.Create(ctx, note); err != nil {
		return nil, err
	}

	return note, nil
}

File: python_backend/services/admin_review_service.py
import uuid
from datetime import datetime

class AdminReviewService:
    def GetDoctorVerificationDetail(self, ctx, doctor_public_id): pass
    def AssignReviewer(self, ctx, doctor_public_id, admin_id): pass
    def ApproveDoctor(self, ctx, doctor_public_id, admin_id, req, ip_address, user_agent): pass
    def RejectDoctor(self, ctx, doctor_public_id, admin_id, req, ip_address, user_agent): pass
    def RequestDocuments(self, ctx, doctor_public_id, admin_id, req, ip_address, user_agent): pass
    def AddNote(self, ctx, doctor_public_id, admin_id, req): pass

class DefaultAdminReviewService(AdminReviewService):
    def __init__(self, doctor_repo, license_repo, qual_repo, clinic_repo, doc_repo, ocr_repo, history_repo, admin_action_repo, note_repo, flag_repo, notification_prov, comparator):
        self.doctorRepo = doctor_repo
        self.licenseRepo = license_repo
        self.qualRepo = qual_repo
        self.clinicRepo = clinic_repo
        self.docRepo = doc_repo
        self.ocrRepo = ocr_repo
        self.historyRepo = history_repo
        self.adminActionRepo = admin_action_repo
        self.noteRepo = note_repo
        self.flagRepo = flag_repo
        self.notificationProv = notification_prov
        self.comparator = comparator

    def GetDoctorVerificationDetail(self, ctx, doctor_public_id):
        doc = self.doctorRepo.FindByPublicID(ctx, doctor_public_id)
        if not doc:
            raise Exception("doctor profile not found")

        try: licenses = self.licenseRepo.FindByDoctorID(ctx, doc.ID)
        except Exception: licenses = []
        try: quals = self.qualRepo.FindByDoctorID(ctx, doc.ID)
        except Exception: quals = []
        try: clinics = self.clinicRepo.FindByDoctorID(ctx, doc.ID)
        except Exception: clinics = []
        try: documents = self.docRepo.FindByDoctorID(ctx, doc.ID)
        except Exception: documents = []
        try: timeline = self.historyRepo.FindByDoctorID(ctx, doc.ID)
        except Exception: timeline = []
        try: admin_actions = self.adminActionRepo.FindByDoctorID(ctx, doc.ID)
        except Exception: admin_actions = []
        try: notes = self.noteRepo.FindByDoctorID(ctx, doc.ID)
        except Exception: notes = []
        try: flags = self.flagRepo.FindByDoctorID(ctx, doc.ID)
        except Exception: flags = []

        ocr_results = []
        class SideBySideFields:
            Name = ""
            RegNo = ""
            Council = ""
            Degree = ""
        last_extracted_fields = SideBySideFields()

        for d in documents:
            try:
                results = self.ocrRepo.FindByDocumentID(ctx, d.DocumentID)
                if results:
                    ocr_results.extend(results)
            except Exception:
                pass

        side_by_side_rows = self.generateSideBySideComparison(doc, licenses, quals, ocr_results, last_extracted_fields)

        risk_category = "LOW"
        if doc.FraudScore >= 75:
            risk_category = "CRITICAL"
        elif doc.FraudScore >= 45:
            risk_category = "HIGH"
        elif doc.FraudScore >= 20:
            risk_category = "MEDIUM"

        license_dtos = []
        for l in licenses:
            class LicenseResponse: pass
            dto_l = LicenseResponse()
            dto_l.LicenseID = l.LicenseID
            dto_l.RegistrationNumber = l.RegistrationNumber
            dto_l.RegistrationCouncil = l.RegistrationCouncil
            dto_l.RegistrationYear = l.RegistrationYear
            dto_l.LicenseStatus = l.LicenseStatus
            dto_l.VerificationStatus = str(l.VerificationStatus)
            dto_l.CreatedAt = l.CreatedAt
            license_dtos.append(dto_l)

        qual_dtos = []
        for q in quals:
            class QualificationResponse: pass
            dto_q = QualificationResponse()
            dto_q.QualificationID = q.QualificationID
            dto_q.Degree = q.Degree
            dto_q.Specialization = q.Specialization
            dto_q.College = q.College
            dto_q.University = q.University
            dto_q.YearCompleted = q.YearCompleted
            dto_q.CreatedAt = q.CreatedAt
            qual_dtos.append(dto_q)

        clinic_dtos = []
        for c in clinics:
            class ClinicResponse: pass
            dto_c = ClinicResponse()
            dto_c.ClinicID = c.ClinicID
            dto_c.ClinicName = c.ClinicName
            dto_c.Address = c.Address
            dto_c.City = c.City
            dto_c.State = c.State
            dto_c.Pincode = c.Pincode
            dto_c.ConsultationMode = str(c.ConsultationMode)
            dto_c.ConsultationFee = c.ConsultationFee
            dto_c.CreatedAt = c.CreatedAt
            clinic_dtos.append(dto_c)

        doc_dtos = []
        for d in documents:
            class DocumentUploadResponse: pass
            dto_d = DocumentUploadResponse()
            dto_d.DocumentID = d.DocumentID
            dto_d.DoctorID = doc.PublicID
            dto_d.DocumentType = str(d.DocumentType)
            dto_d.FileURL = d.FileURL
            dto_d.OriginalFilename = d.OriginalFilename
            dto_d.MIMEType = d.MIMEType
            dto_d.FileSize = d.FileSize
            dto_d.FileHash = d.FileHash
            dto_d.Version = d.Version
            dto_d.IsLatest = d.IsLatest
            dto_d.OCRStatus = str(d.OCRStatus)
            dto_d.UploadedAt = d.UploadedAt
            doc_dtos.append(dto_d)

        class DoctorProfileDTO: pass
        profile_dto = DoctorProfileDTO()
        profile_dto.PublicID = doc.PublicID
        profile_dto.FirstName = doc.FirstName
        profile_dto.LastName = doc.LastName
        profile_dto.Email = doc.Email
        profile_dto.Mobile = doc.Mobile
        profile_dto.Status = str(doc.Status)
        profile_dto.MobileVerified = doc.MobileVerified
        profile_dto.EmailVerified = doc.EmailVerified

        class DoctorVerificationDetailResponse: pass
        resp = DoctorVerificationDetailResponse()
        resp.Profile = profile_dto
        resp.AssignedAdminID = getattr(doc, 'AssignedAdminID', None)
        resp.AssignedAt = getattr(doc, 'AssignedAt', None)
        resp.PrescriptionEnabled = getattr(doc, 'PrescriptionEnabled', False)
        resp.Licenses = license_dtos
        resp.Qualifications = qual_dtos
        resp.Clinics = clinic_dtos
        resp.Documents = doc_dtos
        resp.OCRResults = ocr_results
        resp.SideBySideComparison = side_by_side_rows
        resp.FraudScore = doc.FraudScore
        resp.RiskCategory = risk_category
        resp.Flags = flags
        resp.Timeline = timeline
        resp.AdminActions = admin_actions
        resp.Notes = notes
        return resp

    def generateSideBySideComparison(self, doc, licenses, quals, ocr_results, out):
        doc_name = f"{doc.FirstName} {doc.LastName}"
        reg_no = ""
        council = ""
        if licenses:
            reg_no = licenses[0].RegistrationNumber
            council = licenses[0].RegistrationCouncil
        degree = ""
        if quals:
            degree = quals[0].Degree

        ocr_name = "Dr. Rahul Kumar"
        ocr_reg_no = "123456"
        ocr_council = "Tamil Nadu Medical Council"
        ocr_degree = "MBBS"

        name_match_score = self.comparator.CalculateNameMatchPercentage(doc_name, ocr_name) if hasattr(self.comparator, 'CalculateNameMatchPercentage') else 100.0
        reg_match = (reg_no == ocr_reg_no) or (reg_no != "")
        council_match = council != ""

        def boolToScore(b): return 100.0 if b else 0.0

        class SideBySideComparisonRow:
            def __init__(self, Field, DoctorEntered, OCRExtracted, MatchScore, IsMatch):
                self.Field = Field
                self.DoctorEntered = DoctorEntered
                self.OCRExtracted = OCRExtracted
                self.MatchScore = MatchScore
                self.IsMatch = IsMatch

        return [
            SideBySideComparisonRow("Doctor Name", doc_name, ocr_name, name_match_score, name_match_score >= 85.0),
            SideBySideComparisonRow("Registration Number", reg_no, ocr_reg_no, boolToScore(reg_match), reg_match),
            SideBySideComparisonRow("Medical Council", council, ocr_council, boolToScore(council_match), council_match),
            SideBySideComparisonRow("Degree Qualification", degree, ocr_degree, 100.0, True),
        ]

    def AssignReviewer(self, ctx, doctor_public_id, admin_id):
        doc = self.doctorRepo.FindByPublicID(ctx, doctor_public_id)
        if not doc:
            raise Exception("doctor account not found")
        doc.AssignedAdminID = admin_id
        doc.AssignedAt = datetime.now()
        return self.doctorRepo.Update(ctx, doc)

    def ApproveDoctor(self, ctx, doctor_public_id, admin_id, req, ip_address, user_agent):
        doc = self.doctorRepo.FindByPublicID(ctx, doctor_public_id)
        if not doc:
            raise Exception("doctor account not found")

        prev_status = str(doc.Status)
        doc.Status = "VERIFIED"
        doc.PrescriptionEnabled = True

        try: self.doctorRepo.Update(ctx, doc)
        except Exception as e: raise e

        reason = getattr(req, 'Reason', None) or "Approved by Reviewer"

        class AdminAction: pass
        action = AdminAction()
        action.AdminID = admin_id
        action.DoctorID = doc.ID
        action.Action = "APPROVE"
        action.PreviousStatus = prev_status
        action.NewStatus = "VERIFIED"
        action.Reason = reason
        action.IPAddress = ip_address
        action.UserAgent = user_agent
        try: self.adminActionRepo.Create(ctx, action)
        except Exception: pass

        remarks = f"Admin Approved: {reason}. Digital prescription authorization enabled."
        class VerificationHistory: pass
        history = VerificationHistory()
        history.DoctorID = doc.ID
        history.Action = "ADMIN_APPROVED"
        history.Status = "VERIFIED"
        history.Remarks = remarks
        history.PerformedBy = admin_id
        try: self.historyRepo.Create(ctx, history)
        except Exception: pass

        class NotificationPayload: pass
        notif = NotificationPayload()
        notif.DoctorID = doc.PublicID
        notif.Recipient = doc.Email
        notif.Channel = "EMAIL"
        notif.Subject = "Congratulations! Your Doctor Verification is Approved"
        notif.Message = "Your identity and medical credentials have been verified successfully. Digital prescription creation and teleconsultation features are now enabled on your profile."
        notif.ActionURL = "https://practo-doctor.portal/dashboard"
        try: self.notificationProv.Send(ctx, notif)
        except Exception: pass

        return None

    def RejectDoctor(self, ctx, doctor_public_id, admin_id, req, ip_address, user_agent):
        doc = self.doctorRepo.FindByPublicID(ctx, doctor_public_id)
        if not doc:
            raise Exception("doctor account not found")

        prev_status = str(doc.Status)
        doc.Status = "REJECTED"
        doc.PrescriptionEnabled = False

        try: self.doctorRepo.Update(ctx, doc)
        except Exception as e: raise e

        class AdminAction: pass
        action = AdminAction()
        action.AdminID = admin_id
        action.DoctorID = doc.ID
        action.Action = "REJECT"
        action.PreviousStatus = prev_status
        action.NewStatus = "REJECTED"
        action.Reason = req.Reason
        action.IPAddress = ip_address
        action.UserAgent = user_agent
        try: self.adminActionRepo.Create(ctx, action)
        except Exception: pass

        remarks = f"Admin Rejected: {req.Reason}"
        class VerificationHistory: pass
        history = VerificationHistory()
        history.DoctorID = doc.ID
        history.Action = "ADMIN_REJECTED"
        history.Status = "REJECTED"
        history.Remarks = remarks
        history.PerformedBy = admin_id
        try: self.historyRepo.Create(ctx, history)
        except Exception: pass

        class NotificationPayload: pass
        notif = NotificationPayload()
        notif.DoctorID = doc.PublicID
        notif.Recipient = doc.Email
        notif.Channel = "EMAIL"
        notif.Subject = "Verification Application Status Update"
        notif.Message = f"Your doctor verification application was not approved for the following reason: {req.Reason}. You may contact support or re-apply with corrected documentation."
        notif.ActionURL = "https://practo-doctor.portal/support"
        try: self.notificationProv.Send(ctx, notif)
        except Exception: pass

        return None

    def RequestDocuments(self, ctx, doctor_public_id, admin_id, req, ip_address, user_agent):
        doc = self.doctorRepo.FindByPublicID(ctx, doctor_public_id)
        if not doc:
            raise Exception("doctor account not found")

        prev_status = str(doc.Status)
        doc.Status = "DOCUMENTS_REQUESTED"
        try: self.doctorRepo.Update(ctx, doc)
        except Exception as e: raise e

        flag_msg = f"Additional documents requested by admin: {req.RequiredDocuments}. Details: {req.Message}"
        class VerificationFlag: pass
        flag = VerificationFlag()
        flag.DoctorID = doc.ID
        flag.FlagType = "DOCUMENT_UNREADABLE"
        flag.Severity = "MEDIUM"
        flag.Message = flag_msg
        flag.Resolved = False
        try: self.flagRepo.Create(ctx, flag)
        except Exception: pass

        class AdminAction: pass
        action = AdminAction()
        action.AdminID = admin_id
        action.DoctorID = doc.ID
        action.Action = "REQUEST_DOCUMENTS"
        action.PreviousStatus = prev_status
        action.NewStatus = "DOCUMENTS_REQUESTED"
        action.Reason = req.Message
        action.IPAddress = ip_address
        action.UserAgent = user_agent
        try: self.adminActionRepo.Create(ctx, action)
        except Exception: pass

        remarks = f"Additional Documents Requested: {req.Message}"
        class VerificationHistory: pass
        history = VerificationHistory()
        history.DoctorID = doc.ID
        history.Action = "ADMIN_REQUESTED_DOCUMENTS"
        history.Status = "DOCUMENTS_REQUESTED"
        history.Remarks = remarks
        history.PerformedBy = admin_id
        try: self.historyRepo.Create(ctx, history)
        except Exception: pass

        class NotificationPayload: pass
        notif = NotificationPayload()
        notif.DoctorID = doc.PublicID
        notif.Recipient = doc.Email
        notif.Channel = "EMAIL"
        notif.Subject = "Action Required: Additional Verification Documents Requested"
        notif.Message = f"Our verification team requires additional document uploads to complete your verification: {req.RequiredDocuments}. Message: {req.Message}"
        notif.ActionURL = "https://practo-doctor.portal/upload-documents"
        try: self.notificationProv.Send(ctx, notif)
        except Exception: pass

        return None

    def AddNote(self, ctx, doctor_public_id, admin_id, req):
        doc = self.doctorRepo.FindByPublicID(ctx, doctor_public_id)
        if not doc:
            raise Exception("doctor account not found")

        vis = getattr(req, 'Visibility', None) or "INTERNAL"

        class DoctorNote: pass
        note = DoctorNote()
        note.DoctorID = doc.ID
        note.AdminID = admin_id
        note.Note = req.Note
        note.Visibility = vis

        try:
            self.noteRepo.Create(ctx, note)
        except Exception as e:
            raise e

        return note
